ree/reeEstimateError.cc

```diff
@@ -0,0 +1,458 @@
+	apf::Field* f, mth::Vector<double> const error_dofs)
+{
+	double error = 0.0;
+
+  apf::FieldShape* fs = f->getShape();
+  int type = mesh->getType(e);
+	PCU_ALWAYS_ASSERT(type == apf::Mesh::TET);
+  int nd = apf::countElementNodes(fs, type);
+  int dim = apf::getDimension(mesh, e);
+  double w;
+
+  apf::MeshElement* me = apf::createMeshElement(mesh, e);
+  apf::Element* el = apf::createElement(f, me);
+  int order = 2 * fs->getOrder();
+  int np = apf::countIntPoints(me, order);
+
+  apf::NewArray<apf::Vector3> vectorshape(nd);
+
+  apf::Vector3 p;
+  for (int i = 0; i < np; i++) {
+    apf::getIntPoint(me, order, i, p);
+    double weight = apf::getIntWeight(me, order, i);
+    apf::Matrix3x3 J;
+    apf::getJacobian(me, p, J);
+    double jdet = apf::getJacobianDeterminant(J, dim);
+    w = weight * jdet;
+
+    apf::getVectorShapeValues(el, p, vectorshape);
+    mth::Matrix<double> vectorShape (nd, dim);
+    for (int j = 0; j < nd; j++)
+      for (int k = 0; k < dim; k++)
+        vectorShape(j,k) = vectorshape[j][k];
+
+    mth::Matrix<double> vectorShapeT (dim, nd);
+    mth::transpose(vectorShape, vectorShapeT);
+
+    mth::Vector<double> err_func;
+    mth::multiply(vectorShapeT, error_dofs, err_func);
+
+    error += w * (err_func * err_func);
+  }
+  apf::destroyElement(el);
+  apf::destroyMeshElement(me);
+
+	if (error < 0.0)
+		error = -error;
+
+	return sqrt(error);
+}
+
+apf::Field* computeErrorField(apf::Field* ef, apf::Field* correctedFlux)
+{
+
+	// 1. Create per-element SCALAR error field
+  apf::Field* error_field = apf::createIPField(
+		apf::getMesh(ef), "residual_error_field", apf::SCALAR, 1);
+
+  // 2. Create p+1 order tet ND field
+  int order = ef->getShape()->getOrder();
+  int orderp1 = order+1;
+  apf::Field* efp1 = apf::createField(apf::getMesh(ef),
+	 "orderp1_nedelec_field", apf::SCALAR, apf::getNedelec(orderp1));
+  apf::zeroField(efp1);
+
+  // 2. iterate over all elements of the mesh
+  apf::MeshEntity* el;
+  apf::MeshIterator* it = apf::getMesh(ef)->begin(3);
+  while ((el = apf::getMesh(ef)->iterate(it))) {
+
+    // 2(a). Assemble LHS element matrix
+    mth::Matrix<double> A;
+    assembleElementMatrix( apf::getMesh(ef), el, efp1, A);
+
+    // 2(b). Compute Bilinear Form Vector
+    mth::Vector<double> blf;
+    computeResidualBLF(apf::getMesh(efp1), el, ef, efp1, blf);
+
+    // 2(c). Compute Linear Form Vector
+    mth::Vector<double> lf;
+    assembleDomainLFElementVector(apf::getMesh(efp1), el, efp1, lf);
+
+    // 2(d). Compute Lambda Vector
+    mth::Vector<double> lambda;
+    computeLambdaVector(
+        apf::getMesh(ef), el, ef, efp1, correctedFlux, lambda);
+
+    // 2(e). Assemble RHS element vector = blf - lf - lambda
+    mth::Vector<double> B(blf.size());
+    B.zero();
+    B += blf; B -= lf; B -= lambda;
+
+    // 2(f). Get List of Essential Dofs
+    apf::NewArray<int> ess_dofs, uness_dofs;
+    getEssentialElementNDDofs(
+        apf::getMesh(efp1), el, efp1, ess_dofs, uness_dofs);
```

Comment:
Several indentation issues in this section of code.

Reply:
Done.